### PR TITLE
[sk] Improved Configuration Framework

### PR DESCRIPTION
**Pull request by skunichetty**
# Summary
Introduces a more structured configuration framework, enabling users to have a more robust method to handle authentication features for data IO. Two new features were added:
- A new configuration framework based around `BaseConfigLoader`, a base class that requires all loader to have the following two functions
    - `get` - get the configuration setting given the respective key. If setting doesn't exist, returns None.
    - `contains` - check if the configuration setting corresponding to the given key exists
- A new unified set of keys that define all the secrets used by data loaders (stored under the `ConfigKey` enum). The following template shows how a configuration file with these new keys look:
```yaml
default:
  # Default profile created for data IO access.
  # Add your credentials for the source you use, and delete the rest.
  AWS_ACCESS_KEY_ID = AWS Access Key ID credential
  AWS_SECRET_ACCESS_KEY = AWS Secret Access Key credential
  AWS_SESSION_TOKEN = AWS Session Token (used to generate temp DB credentials)
  AWS_REGION = AWS Region
  GOOGLE_SERVICE_ACC_KEY = Service account key (add nested object here)
  GOOGLE_SERVICE_ACC_KEY_FILEPATH = Path to service account key
  POSTGRES_DBNAME = Database name
  POSTGRES_USER = Database login username
  POSTGRES_PASSWORD = Database login password
  POSTGRES_HOST = Database hostname
  POSTGRES_PORT = PostgreSQL database port
  REDSHIFT_DBNAME = Name of Redshift database to connect to
  REDSHIFT_HOST = Redshift Cluster hostname
  REDSHIFT_PORT = Redshift Cluster port. Optional, defaults to 5439
  REDSHIFT_TEMP_CRED_USER = Redshift temp credentials username
  REDSHIFT_TEMP_CRED_PASSWORD = Redshift temp credentials password
  REDSHIFT_DBUSER = Redshift database user to generate credentials for
  REDSHIFT_CLUSTER_ID = Redshift cluster ID
  REDSHIFT_IAM_PROFILE = Name of the IAM profile to generate temp credentials with
  SNOWFLAKE_USER = Snowflake username
  SNOWFLAKE_PASSWORD = Snowflake password
  SNOWFLAKE_ACCOUNT = Snowflake account ID (including region)
  SNOWFLAKE_DEFAULT_WH = Default warehouse to use. Optional
  SNOWFLAKE_DEFAULT_DB = Default database to use. Optional
  SNOWFLAKE_DEFAULT_SCHEMA = Default schema to use. Optional

```

Currently, support for loading from configuration file, environment variables, and AWS Secrets Manager is supported.

## Notes
1. I'll push integration with the old configuration system and data IO clients in a separate PR (this one is already long enough)
2. This PR is mainly unit tests and documentation - the amount of new code added is quite small.

# Tests
Unit tests were written to test the behavior of the configuration loaders.

cc:
@wangxiaoyou1993 @dy46

### CODE REVIEW

**Comment by wangxiaoyou1993** [comment]
Can we use combination of the environment variables and plain text in the config yaml?

**Comment by wangxiaoyou1993** [comment]
I wonder we can support something like this: https://docs.getdbt.com/reference/dbt-jinja-functions/env_var

File: mage_ai/io/config.py [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+
+        Returns:
+            Any: The configuration setting stored under `env_var`
+        """
+        return os.getenv(env_var)
+
+
+class ConfigFileLoader(BaseConfigLoader):
+    def __init__(
+        self, filepath: os.PathLike = './default_repo/io_config.yaml', profile='default'
```
Comment (wangxiaoyou1993) [inline mage_ai/io/config.py:197]: do not use default_repo here. default_repo is only for development. use `get_repo_path()` instead.

**Review by wangxiaoyou1993 — APPROVED** [review]

File: mage_ai/io/config.py (hunk 1 above)
Comment (skunichetty) [reply mage_ai/io/config.py]: Fixed.

**Pull request merged**